File: Day #1613/solution.cpp
// Skyline via sweep-line + max-heap (multiset). Emit key point when max height changes.
// Time: O(n log n), Space: O(n).
#include <bits/stdc++.h>
using namespace std;

int main() {
    vector<array<int,3>> buildings = {{0,15,3},{4,11,5},{19,23,4}};
    vector<pair<int,int>> events; // (x, signed height): start = -h, end = +h
    for (auto& b : buildings) {
        events.push_back({b[0], -b[2]});
        events.push_back({b[1],  b[2]});
    }
    sort(events.begin(), events.end());

    multiset<int> heights = {0};
    int prev = 0;
    vector<pair<int,int>> res;
    for (auto& e : events) {
        int x = e.first, h = e.second;
        if (h < 0) heights.insert(-h);
        else       heights.erase(heights.find(h));
        int cur = *heights.rbegin();
        if (cur != prev) {
            res.push_back({x, cur});
            prev = cur;
        }
    }

    cout << "[";
    for (size_t i = 0; i < res.size(); ++i) {
        cout << "(" << res[i].first << ", " << res[i].second << ")";
        if (i + 1 < res.size()) cout << ", ";
    }
    cout << "]" << endl;
    return 0;
}

File: Day #1613/solution.rs
// Skyline via sweep-line + BTreeMap multiset of active heights. Emit point when max changes.
// Time: O(n log n), Space: O(n).
use std::collections::BTreeMap;

fn get_skyline(buildings: &[(i32, i32, i32)]) -> Vec<(i32, i32)> {
    // events: (x, signed height) with start = -h, end = +h
    let mut events: Vec<(i32, i32)> = Vec::new();
    for &(left, right, h) in buildings {
        events.push((left, -h));
        events.push((right, h));
    }
    events.sort();

    let mut heights: BTreeMap<i32, i32> = BTreeMap::new();
    heights.insert(0, 1);
    let mut prev = 0;
    let mut res: Vec<(i32, i32)> = Vec::new();
    for &(x, h) in &events {
        if h < 0 {
            *heights.entry(-h).or_insert(0) += 1;
        } else {
            if let Some(c) = heights.get_mut(&h) {
                *c -= 1;
                if *c == 0 {
                    heights.remove(&h);
                }
            }
        }
        let cur = *heights.keys().next_back().unwrap();
        if cur != prev {
            res.push((x, cur));
            prev = cur;
        }
    }
    res
}

fn main() {
    let buildings = [(0, 15, 3), (4, 11, 5), (19, 23, 4)];
    let res = get_skyline(&buildings);
    let parts: Vec<String> = res.iter().map(|&(x, h)| format!("({}, {})", x, h)).collect();
    println!("[{}]", parts.join(", "));
}
